spec/ruby/core/io/timeout_spec.rb

```diff
@@ -0,0 +1,159 @@
+# -*- encoding: utf-8 -*-
+require_relative '../../spec_helper'
+
+ruby_version_is "3.2" do
+  describe "IO#timeout" do
+    before :each do
+      @fname = tmp("io_timeout.txt")
+      @file = File.open(@fname, "a+")
+
+      @rpipe, @wpipe = IO.pipe
+      # There is no strict long term standard for pipe limits (2**16 bytes currently). This is an attempt to set a safe
+      # enough size to test a full pipe.
+      @more_than_pipe_limit = 1 << 18
+    end
+
+    after :each do
+      @rpipe.close
+      @wpipe.close
+
+      @file.close
+      rm_r @fname
+    end
+
+    it "files have timeout attribute" do
+      @fname = tmp("io_timeout_attribute.txt")
+      touch(@fname)
+
+      @file.timeout.should == nil
+
+      @file.timeout = 1.23
+      @file.timeout.should == 1.23
+    end
+
+    it "IO instances have timeout attribute" do
+      @rpipe.timeout.should == nil
+      @wpipe.timeout.should == nil
+
+      @rpipe.timeout = 1.23
+      @wpipe.timeout = 4.56
+
+      @rpipe.timeout.should == 1.23
+      @wpipe.timeout.should == 4.56
+    end
+
+    it "raises IO::TimeoutError when timeout is exceeded for .read" do
+      @rpipe.timeout = 0.001
+      -> { @rpipe.read.should }.should raise_error(IO::TimeoutError)
+    end
+
+    it "raises IO::TimeoutError when timeout is exceeded for .read(n)" do
+      @rpipe.timeout = 0.001
+      -> { @rpipe.read(3) }.should raise_error(IO::TimeoutError)
+    end
+
+    it "raises IO::TimeoutError when timeout is exceeded for .gets" do
+      @rpipe.timeout = 0.001
+      -> { @rpipe.gets }.should raise_error(IO::TimeoutError)
+    end
+
+    it "makes a last attempt to read when the time is over" do
+      @wpipe.write("abc")
+      @wpipe.close
+
+      @rpipe.timeout = 0
+      @rpipe.read(3).should == "abc"
+    end
+
+    it "waits for a read and completes" do
+      @rpipe.timeout = 0.2
+      can_write = false
+
+      t = Thread.new do
+        Thread.pass until can_write
+
+        sleep(0.1)
+        @wpipe.write("abc")
+        @wpipe.close
+      end
+
+      can_write = true
+      @rpipe.read(3).should == "abc"
+
+      t.join
+    end
```

Comment:
This feels very brittle, and also 200ms for one spec is very slow.

I think you could use Thread#status/stop? to achieve the same without sleep.
The thread `t` could `Thread.pass until Thread.main.stop?`, which means it's blocked in read, then just write to the pipe.